Add l1 and l1+l2 regularizers for PyTorch backend

vl-dud opened the pull request:
L1 regularizer is added. Not quite sure about the line:
```
l1_loss = torch.sum(torch.stack([torch.sum(p.abs()) for p in trainable_variables]))
```
Perhaps regularization should not be applied to `self.external_trainable_variables`? Then it must be like:
```
l1_loss = torch.sum(torch.stack([torch.sum(p.abs()) for p in self.net.parameters()]))
```
However, in current l2 regularization `trainable_variables` (both `self.net.parameters()` and `self.external_trainable_variables`) are used.

lululxvi commented:
You are right that regularization should not be applied to `self.external_trainable_variables`. Let us fix this first in this PR.

vl-dud commented:
Fixed. Please note that I replaced
```
raise ValueError("L-BFGS optimizer doesn't support weight_decay > 0")
```
with
```
print(f"Warning: weight decay is ignored for {self.opt_name}")
```

And I removed the exception
```
raise ValueError("AdamW optimizer requires non-zero weight decay")
```
because AdamW will still start with the default `weight_decay`

lululxvi commented:
Let us fix l2 in this PR. Use a new PR for adding l1 and l1+l2.

vl-dud commented:
Take a look now please.

lululxvi commented on deepxde/model.py:
```diff
@@ -330,30 +330,33 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
-                    weight_decay=self.net.regularizer[1],
-                )
+                if self.opt_name in ["L-BFGS", "L-BFGS-B"]:
+                    print(f"Warning: weight decay is ignored for {self.opt_name}")
```
raise ValueError, not only warning

vl-dud replied:
Fixed

lululxvi commented on deepxde/optimizers/pytorch/optimizers.py:
```diff
@@ -33,21 +31,13 @@ def get(params, optimizer, learning_rate=None, decay=None, weight_decay=0):
-            )
+            optim = torch.optim.Adam(params, lr=learning_rate)
         elif optimizer == "adamw":
-            if weight_decay == 0:
```
We need to keep this error to avoid any unexcepted effect.

vl-dud replied:
Fixed

lululxvi commented on deepxde/optimizers/pytorch/optimizers.py:
```diff
@@ -9,14 +9,12 @@ def is_external_optimizer(optimizer):
     return optimizer in ["L-BFGS", "L-BFGS-B"]
 
 
-def get(params, optimizer, learning_rate=None, decay=None, weight_decay=0):
```
Keep this `weight_decay=0`. Then use the following in model.py
```
optimizer_params = [
                {"params": self.net.parameters()},
                {"params": self.external_trainable_variables, "weight_decay": 0},
            ]
``

vl-dud replied:
Fixed.

lululxvi commented:
In line 113, emphasize that `external_trainable_variables` will not apply regularization.

lululxvi commented on deepxde/model.py:
```diff
@@ -330,30 +330,32 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
+        if self.opt_name in ["L-BFGS", "L-BFGS-B"]:
+            optimizer_params = (
+                list(self.net.parameters()) + self.external_trainable_variables
+            )
```
Add a warning here that external_trainable_variables will also happy regularization if weight decay > 0.

vl-dud replied:
If weight decay > 0, an exception is thrown in *optimizers.py*:
```
elif optimizer in ["L-BFGS", "L-BFGS-B"]:
	if weight_decay > 0:
		raise ValueError("L-BFGS optimizer doesn't support weight_decay > 0")
```
The warning doesn't seem to make much sense, does it?

lululxvi replied:
I still suggest to add this warning, as in the future, L-BFGS may support weight decay > 0.

vl-dud commented:
> In line 113, emphasize that `external_trainable_variables` will not apply regularization.

Fixed.

lululxvi commented on deepxde/model.py:
```diff
@@ -112,9 +112,11 @@ def compile(
+                applied to these variables. If the backend is tensorflow.compat.v1,
+                `external_trainable_variables` is ignored, and all trainable ``dde.Variable``
+                objects are automatically collected.
+
```
Delete this empty line.

vl-dud replied:
Fixed.

lululxvi commented on deepxde/model.py:
```diff
@@ -330,30 +331,37 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
+        else:
+            optimizer_params = [
+                {"params": self.net.parameters()},
+                {"params": self.external_trainable_variables, "weight_decay": 0},
```
What if external_trainable_variables = None?

vl-dud replied:
Although the PyTorch optimizer handles empty values correctly, you're right — it’s better to make this point more explicit. Fixed.

lululxvi commented on deepxde/model.py:
```diff
@@ -330,30 +331,39 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
+        optimizer_params = self.net.parameters()
+        if self.external_trainable_variables:
+            if self.opt_name in ["L-BFGS", "L-BFGS-B"]:
+                optimizer_params = (
```
Add a comment:
```
L-BFGS doesn't support per-parameter options.
```

vl-dud replied:
Fixed
